pkg/approver/aws/csr.go

```diff
@@ -0,0 +1,285 @@
+		return nil
+	}
+
+	x509cr, err := csrutil.ParseCSR(csr)
+	if err != nil {
+		return err
+	}
+
+	glog.V(4).Infof("handle: running recognizers on %s", csr.GetName())
+	csrrs := ar.recognizers()
+	for _, csrr := range csrrs {
+		rs := csrr.recognizers
+		approved := true
+		for _, r := range rs {
+			if rerr := r(csr, x509cr); err != nil {
+				glog.V(4).Infof("handle: %v", rerr)
+				approved = false
+				break
+			}
+		}
+		if !approved {
+			continue
+		}
+
+		glog.V(4).Infof("csr %s was approved! message: %s", csr.GetName(), csrr.successMessage)
+		csr.Status.Conditions = append(csr.Status.Conditions, certificates.CertificateSigningRequestCondition{
+			Type:    certificates.CertificateApproved,
+			Reason:  "AutoApproved",
+			Message: csrr.successMessage,
+		})
+		_, err = ar.kubeClient.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(csr)
+		if err != nil {
+			return fmt.Errorf("error updating approval for csr: %v", err)
+		}
+		break
+	}
+
+	return nil
+}
+
+type recognizerFunc func(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) error
+type csrRecognizer struct {
+	recognizers    []recognizerFunc
+	successMessage string
+}
+
+func (ar *Approver) recognizers() []csrRecognizer {
+	return []csrRecognizer{{
+		// ensures: node present their client cert, exist(running) in cluster, belong to whitelist of ASGs.
+		recognizers:    []recognizerFunc{isSelfNodeClientCert, ar.isValidNode(ar.aws.instanceID), ar.isValidASG(ar.aws.autoScalingGroupID)},
+		successMessage: "kube-aws-approver approved self node client cert",
+	}, {
+		// ensures: node username has instance-id, presented instance-id match instance-id by aws for node name in running state, doesn't belong to cluster, belong to whitelist of ASGs.
+		recognizers:    []recognizerFunc{isNodeClientCert, ar.isValidNewNode(ar.aws.instanceID), ar.isValidASG(ar.aws.autoScalingGroupID)},
+		successMessage: "kube-aws-approver approved new node client cert",
+	}}
+}
+
+// isNodeClientCert checks if
+// - x509.CertificateRequest Organization is equal to system:nodes
+// - x509.CertificateRequest DNSNames list is empty
+// - x509.CertificateRequest EmailAddresses list is empty
+// - x509.CertificateRequest IPAddresses list is empty
+// - x509.CertificateRequest CommonName has 'system:node:' prefix
+// - CertificateSigningRequest Username has 'system:bootstrappers:' prefix
+// - CertificateSigningRequest has valid client key usages. (https://github.com/kubernetes/client-go/blob/7cd1d3291b7d9b1e2d54d4b69eb65995eaf8888e/util/certificate/csr/csr.go#L66-L70)
+// upstream: https://github.com/kubernetes/kubernetes/blob/7488d1c9210e60aef9ad49f07cb5d8a24152db88/pkg/controller/certificates/approver/sarapprove.go#L179
+// Returns nil when conditions met.
+func isNodeClientCert(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) error {
+	if !reflect.DeepEqual([]string{"system:nodes"}, x509cr.Subject.Organization) {
```

Comment:
nit: never use DeepEqual outside of tests

```
if len(x509cr.Subject.Organization) != 1 || x509cr.Subject.Organization[0] != "system:nodes" {

}
```

Reply:
done